legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java

```diff
@@ -244,47 +317,114 @@ public ConnectionStateEvictionTask(long durationInSeconds)
         public void run()
         {
             ConnectionStateManager instance = ConnectionStateManager.getInstance();
-            instance.purge(durationInSeconds);
+            try
+            {
+                instance.purge(durationInSeconds);
+            }
+            catch (Exception e)
+            {
+                LOGGER.error("Datasource Connection State purge failed {}", e);
+            }
         }
     }
 
-    public static class ConnectionStatePOJO
+    public DataSourceWithStatistics getDataSourceForIdentityIfAbsentBuild(IdentityState identityState, DataSourceSpecification dataSourceSpecification, Supplier<DataSource> dataSourceBuilder)
     {
-        String poolName;
-        boolean credentialSupplierPresent;
-        long ageInMillis;
 
-        public ConnectionStatePOJO() {
+        String principal = identityState.getIdentity().getName();
+        String poolName = poolNameFor(identityState.getIdentity(), dataSourceSpecification.getConnectionKey());
+        ConnectionKey connectionKey = dataSourceSpecification.getConnectionKey();
+        //thread safe as concurrent map;
+        Function0<DataSourceWithStatistics> dsSupplier = () -> new DataSourceWithStatistics(poolName,identityState,dataSourceSpecification);
+        DataSource dataSource = this.connectionPools.getIfAbsentPut(poolName, dsSupplier).getDataSource();
+        if (dataSource == null)
+        {
+            synchronized (poolLockManager.getLock(poolName))
+            {
+                dataSource = this.connectionPools.getIfAbsentPut(poolName, dsSupplier).getDataSource();
+                if (dataSource == null)
+                {
+                    LOGGER.info("Pool not found for [{}] for datasource [{}], creating one", principal, connectionKey.shortId());
```

Comment:
:confused: I don't follow this synchronization code. 

Lines 339 and 344 use the map's lock (which is tied to the poolName). In addition line 342 uses an explicit lock also tied to the poolName.

In the previous implementation (when this code was in DataSourceSpecification.java), the locking idiom was 
```
object = map.get(key)
if (object == null)
{
   synchronized(makeLock(key))
  {
     object = map.get(key)
     if (object == null) { map.put(key, makeObject()) } 
  }
}
```
If getIfAbsentPut is atomic, do we need the null checks and the explicit lock ?

Reply:
discussed offline will add more commentary to help developers understand